src/Build/Evaluation/ItemSpec.cs

```diff
@@ -292,6 +307,25 @@ public bool MatchesItem(I item)
             return false;
         }
 
+        /// <summary>
+        ///     Return true if any of the given <paramref name="metadata" /> matches the metadata on <paramref name="item" />
+        /// </summary>
+        /// <param name="item">The item to attempt to find a match for based on matching metadata</param>
+        /// <param name="metadata">Names of metadata to look for matches for</param>
+        /// <returns></returns>
+        public bool MatchesItemOnMetadata(IItem item, ImmutableList<string> metadata)
```

Comment:
You should reuse this in the [intrinsic task remove](https://github.com/microsoft/msbuild/pull/4997/files#diff-0d4693ff9ac2fd56e6fda64d0fd0c6cbR287) as well, to reduce duplication.